tests/topotests/lib/topotest.py

```diff
@@ -66,144 +67,140 @@ def has_errors(self):
+                closest_idx = None
+                for idx1, v1 in zip(range(0, len(d1)), d1):
+                    tmp_diff = gen_json_diff_report(v1, v2, path=add_idx(idx1))
+                    if not has_errors(tmp_diff):
+                        found_match = True
+                        del d1[idx1]
+                        break
+                    elif not closest_diff or get_errors_n(tmp_diff) < get_errors_n(closest_diff):
+                        closest_diff = tmp_diff
+                        closest_idx = idx1
+                if not found_match and isinstance(v2, (list, dict)):
+                    sub_error = "\n\n\t" + "\t".join(get_errors(closest_diff).splitlines(True))
+                    acc = add_error(acc, "d2 has the following element at index " + str(idx2) + " which is not present in d1: \n\n" + dump_json(v2)
+                                          + "\n\n\tClosest match in d1 is at index " + str(closest_idx) + " with the following errors: " + sub_error)
+                if not found_match and not isinstance(v2, (list, dict)):
+                    acc = add_error(acc, "d2 has the following element at index " + str(idx2) + " which is not present in d1: " + dump_json(v2))
+    elif isinstance(d1, dict) and isinstance(d2, dict) and exact:
+        invalid_keys_d1 = [k for k in d1.keys() if k not in d2.keys()]
+        invalid_keys_d2 = [k for k in d2.keys() if k not in d1.keys()]
+        for k in invalid_keys_d1:
+            acc = add_error(acc, "d1 has key '" + str(k) + "' which is not present in d2")
+        for k in invalid_keys_d2:
+            acc = add_error(acc, "d2 has key '" + str(k) + "' which is not present in d1")
+        valid_keys_intersection = [k for k in d1.keys() if k in d2.keys()]
+        for k in valid_keys_intersection:
+            acc = merge_errors(acc, gen_json_diff_report(d1[k], d2[k], exact=exact, path=add_key(k)))
+    elif isinstance(d1, dict) and isinstance(d2, dict):
+        none_keys = [k for k, v in d2.items() if v == None]
+        none_keys_present = [k for k in d1.keys() if k in none_keys]
+        for k in none_keys_present:
+            acc = add_error(acc, "d1 has key '" + str(k) + "' which is not supposed to be present")
+        keys = [k for k, v in d2.items() if v != None]
+        invalid_keys_intersection = [k for k in keys if k not in d1.keys()]
+        for k in invalid_keys_intersection:
+            acc = add_error(acc, "d2 has key '" + str(k) + "' which is not present in d1")
+        valid_keys_intersection = [k for k in keys if k in d1.keys()]
+        for k in valid_keys_intersection:
+            acc = merge_errors(acc, gen_json_diff_report(d1[k], d2[k], exact=exact, path=add_key(k)))
+    else:
+        acc = add_error(acc, "d1 has element of type " + json_type(d1) + " but the corresponding element in d2 is of type " + json_type(d2), points=2)
+
+    return acc
 
-def json_cmp(d1, d2):
+
+def json_cmp(d1, d2, exact=False):
     """
```

Comment:
Can you document `exact` and tell the reader about `'*'` and `'__ordered__'`?

I think the choices you've made for `'*'` and `'__ordered__'` are good. For an example JSON users unofficially recommend `'__comment__'` for commentaries since the language itself does not provide.

Reply:
Done.